src/pip/_internal/req/constructors.py

```diff
@@ -200,6 +200,58 @@ def install_req_from_editable(
     )
 
 
+def _get_path_or_url(path, name):
+    # type: (str, str) -> str
+    """
+    First, it checks whether a provided path is an installable directory
+    (e.g. it has a setup.py). If it is, returns the path.
+
+    If false, check if the path is an archive file (such as a .whl).
+    The function checks if the path is a file. If false, if the path has
+    an @, it will treat it as a PEP 440 URL requirement and return the path.
+    """
+    if os.path.isdir(path) and _looks_like_path(name):
+        if not is_installable_dir(path):
+            raise InstallationError(
+                "Directory %r is not installable. Neither 'setup.py' "
+                "nor 'pyproject.toml' found." % name
+            )
+        return path_to_url(path)
+    elif is_archive_file(path):
+        if os.path.isfile(path):
+            return path_to_url(path)
+        else:
+            urlreq_parts = name.split('@', 1)
+            if len(urlreq_parts) < 2 or _looks_like_path(urlreq_parts[0]):
```

Comment:
Like the above, if you test the opposite condition, you can put the simpler code inside the `if` (namely `return None`).

Reply:
With the return statement I forgot to add, that would give 5 return statements. I'm not sure this would be a better idea.

Reply:
The number doesn’t matter so much, and it’s the same number either way (my suggestion was only about the order). The suggestion was to put the simpler code inside the if (one line in this case). That way the slightly more complicated code can be unindented. This technique is called “early return” or sometimes “guard clause.”

Reply:
Good to know. I'll change it then.